crossdock/client/headers/behavior.go

```diff
@@ -0,0 +1,173 @@
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package headers
+
+import (
+	"time"
+
+	"github.com/yarpc/yarpc-go/crossdock/client/behavior"
+	"github.com/yarpc/yarpc-go/crossdock/client/random"
+	"github.com/yarpc/yarpc-go/crossdock/thrift/echo"
+	"github.com/yarpc/yarpc-go/crossdock/thrift/echo/yarpc/echoclient"
+	"github.com/yarpc/yarpc-go/encoding/json"
+	"github.com/yarpc/yarpc-go/encoding/raw"
+	"github.com/yarpc/yarpc-go/encoding/thrift"
+	"github.com/yarpc/yarpc-go/transport"
+
+	"golang.org/x/net/context"
+)
+
+// Run runs the headers behavior
+func Run(s behavior.Sink, ps behavior.Params) {
+	fatals := behavior.Fatals(s)
+	assert := behavior.Assert(s)
+	checks := behavior.Checks(s)
+
+	rpc := createRPC(s, ps)
+
+	encoding := ps.Param(EncodingParam)
+	fatals.NotEmpty(encoding, "encoding is required")
+
+	var caller headerCaller
+
+	switch encoding {
+	case "raw":
+		caller = rawCaller{raw.New(rpc.Channel("yarpc-test"))}
+	case "json":
+		caller = jsonCaller{json.New(rpc.Channel("yarpc-test"))}
+	case "thrift":
+		caller = thriftCaller{echoclient.New(rpc.Channel("yarpc-test"))}
+	default:
+		fatals.True(false, "unknown encoding %q", encoding)
+	}
+
+	token1 := random.String(10)
+	token2 := random.String(10)
+
+	tests := []struct {
+		desc string
+		give transport.Headers
+		want transport.Headers
+	}{
+		{
+			"valid headers",
+			transport.Headers{"token1": token1, "token2": token2},
+			transport.Headers{"token1": token1, "token2": token2},
+		},
+		{
+			"non-string values",
+			transport.Headers{"token": "42"},
+			transport.Headers{"token": "42"},
+		},
+		{
+			"empty strings",
+			transport.Headers{"token": ""},
+			transport.Headers{"token": ""},
+		},
+		{
+			"no headers",
+			nil,
+			transport.Headers{},
+		},
+		{
+			"empty map",
+			transport.Headers{},
+			transport.Headers{},
+		},
+		{
+			"varying casing",
+			transport.Headers{"ToKeN1": token1, "tOkEn2": token2},
+			transport.Headers{"token1": token1, "token2": token2},
+			// TODO(abg): Do we expect to get back the same casing or
```

Comment:
we can't get back the keys with the same casing right? this seems like the right behaviour to me